File: typescript/packages/bench/src/reports/orderVolumeReport.ts
/** Orders grouped by intake priority. */

import type { ReportRow } from "../contracts/reportContract";
import type { Dataset } from "../dataset";
import { reportRow } from "../contracts/reportContract";
import { Priority } from "@atelier/core";
import { AbstractReport } from "../support/abstractReport";

/** Orders grouped by intake priority. */
export class OrderVolumeReport extends AbstractReport {
    /** Registry slug; also the URL segment. */
    static readonly SLUG = "order-volume";

    constructor() {
        super(OrderVolumeReport.SLUG, "Order volume");
    }

    override rows(data: Dataset): ReportRow[] {
        const counts = new Map<string, number>();
        for (const order of data.orders) {
            const label = Priority.label(order.priority);
            counts.set(label, (counts.get(label) ?? 0) + 1);
        }

        return [...counts.entries()]
            .sort(([left], [right]) => left.localeCompare(right))
            .map(([label, count]) => reportRow(label, count));
    }
}

File: rust/crates/atelier-bench/src/reports/order_volume_report.rs
//! Orders grouped by intake priority.

use crate::contracts::report_contract::{ReportContract, ReportRow};
use crate::dataset::Dataset;
use crate::support::abstract_report::ReportBase;
use itertools::Itertools;

/// Orders grouped by intake priority.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct OrderVolumeReport {
    base: ReportBase,
}

impl OrderVolumeReport {
    /// Registry slug; also the URL segment.
    pub const SLUG: &'static str = "order-volume";

    pub const fn new() -> Self {
        Self {
            base: ReportBase::new(Self::SLUG, "Order volume"),
        }
    }

    /// Shared identity and rendering precision.
    pub const fn base(&self) -> &ReportBase {
        &self.base
    }
}

impl Default for OrderVolumeReport {
    fn default() -> Self {
        Self::new()
    }
}

impl ReportContract for OrderVolumeReport {
    fn slug(&self) -> &'static str {
        self.base.slug()
    }

    fn title(&self) -> &'static str {
        self.base.title()
    }

    fn rows(&self, data: &Dataset) -> Vec<ReportRow> {
        data.orders
            .iter()
            .map(|order| order.priority.label())
            .counts()
            .into_iter()
            .sorted_by_key(|(label, _)| *label)
            .map(|(label, count)| ReportRow::new(label, count as f64))
            .collect()
    }
}
